Add implicit batch experimental

pooyadavoodi opened the pull request:
- Add experimental argument `use_implicit_batch`. This is the first step in adding dynamic shapes to TF-TRT, so it still doesn't work but it provides the scaffolding for the next PRs.
- Adds a simple test for rewriter_config_template which is the only way to use `use_implicit_batch` until we take it out of experimental.
- Improves argument checks in trt_convert.py

Commit:
Add use_implicit_batch to TF-TRT as as experimental feature

Use buildEngineWithConfig from TensorRT 6.0

Enable custom rewriter_config_template

Commit:
Fix some of the pylint errors

aaroey commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -289,6 +292,8 @@ TRTEngineOp::TRTEngineOp(OpKernelConstruction* context)
   }
   OP_REQUIRES_OK(context, context->GetAttr("max_cached_engines_count",
                                            &max_cached_engines_));
+  OP_REQUIRES_OK(context,
```
As mentioned in ops/trt_engine_op.cc, please use implicit attribute and don't use `OP_REQUIRES_OK`, but set it to `true` if the attribute is not set.

pooyadavoodi replied:
Done

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -419,17 +427,27 @@ Status GetTrtBroadcastShape(const TRT_TensorOrWeights& operand_l,
-      std::max(operand_l.GetTrtDims().nbDims + (operand_l.is_tensor() ? 1 : 0),
-               operand_r.GetTrtDims().nbDims + (operand_r.is_tensor() ? 1 : 0));
+      std::max(operand_l.GetTrtDims().nbDims +
+                   (use_implicit_batch && (operand_l.is_tensor() ? 1 : 0)),
```
Should the condition be `(use_implicit_batch && operand_l.is_tensor())`?

pooyadavoodi replied:
I don't think there is a difference in the final result?

I just simplified it to the following.
Let me know if you think this is better:

```
      std::max(operand_l.GetTrtDims().nbDims + 
                   use_implicit_batch && operand_l.is_tensor(),
               operand_r.GetTrtDims().nbDims +
                   use_implicit_batch && operand_r.is_tensor());
```

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -1212,26 +1236,37 @@ static void InitializeTrtPlugins(nvinfer1::ILogger* trt_logger) {
   // Create the network.
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+  const uint32_t flags = use_implicit_batch_ ? 0U :
+      1U << static_cast<int>(
```
Please add () for `1U << ...`.

pooyadavoodi replied:
Done

pooyadavoodi commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -189,6 +191,41 @@ def _check_conversion_params(conversion_params):
+        if trt_optimizer:
+          raise TypeError("Found more than one TensorRTOptimizer in "
+                          "rewriter_config_template while only one is allowed.")
+        trt_optimizer = optim
```
`trt_optimizer` has to be in the template only if the template is set.

By default, the template is not set (i.e. equals None), and trt_optimizer is added by the converter. But users might want to add the trt_optimizer themselves for some reason (e.g. disabling `use_implicit_batch`).

I don't think there is any use for setting the template by users without adding trt_optimizer to it.

pooyadavoodi replied:
Added a comment

Commit:
Add comment for requring trt in rewriter_config_template, optim --> optimizer

Commit:
Use parenthesis instead of breaking line with \

pooyadavoodi commented on tensorflow/compiler/tf2tensorrt/ops/trt_engine_op.cc:
```diff
@@ -41,6 +41,7 @@ REGISTER_OP("TRTEngineOp")
     .Attr("precision_mode: {'FP32', 'FP16', 'INT8'}")
     .Attr("calibration_data: string = ''")
     .Attr("use_calibration: bool = true")
+    .Attr("use_implicit_batch: bool = true")
```
Does it fail even though there is a default value for the attribute?

aaroey replied:
I think it will. It'll be great if you can help to add a test for that as we chatted.

pooyadavoodi replied:
Done

pooyadavoodi commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -189,6 +191,41 @@ def _check_conversion_params(conversion_params):
+  if conversion_params.rewriter_config_template:
+    rewriter_cfg = conversion_params.rewriter_config_template
+    trt_optimizer = None
+    for optim in rewriter_cfg.custom_optimizers:
```
Done

pooyadavoodi commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -260,36 +297,29 @@ def get_tensorrt_rewriter_config(conversion_params, is_v2=False):
+    optimizer.name = "TensorRTOptimizer"
+    optimizer.parameter_map[
+        "minimum_segment_size"].i = conversion_params.minimum_segment_size
+    optimizer.parameter_map["max_workspace_size_bytes"].i = \
```
Done

Commit:
Simplify computing broadcast_num_dims

Commit:
Enclose statement in parentheses

Commit:
Use data member precision_mode_ (fix compilation issue)

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -445,9 +445,9 @@ Status GetTrtBroadcastShape(const TRT_TensorOrWeights& operand_l,
   const int broadcast_num_dims =
       std::max(operand_l.GetTrtDims().nbDims +
-                   (use_implicit_batch && (operand_l.is_tensor() ? 1 : 0)),
+                   use_implicit_batch && operand_l.is_tensor(),
```
Please add () to make it clear.

pooyadavoodi replied:
Done

Commit:
Fix bug: use_calibration --> use_calibration_

Commit:
Add script to generate TF-TRT model to be used for testing

Commit:
Add tftrt_2.0_saved_model

This model can be loaded and executed in tests to ensure
backwards compatibility across TF versions.

Commit:
Add backward compatibility test

Commit:
Move data of quantization_mnist_test to testdata/mnist

Commit:
Add mnist dir, add correctness check for bwd compatibility test

Commit:
Fix pylint for backwards compatibility test

Commit:
Sort targets in BUILD and shorten statement for pylint

Commit:
Make use_implicit_batch an implicit attribute

Set use_implicit_batch to true in case it is not present in graph

Add parenthesese for readability

gbaned commented:
@pooyadavoodi Can you please resolve conflicts? Thanks!

aaroey commented:
Also @sanjoy

pooyadavoodi commented:
> @pooyadavoodi Can you please resolve conflicts? Thanks!

Done

I've added the commits from https://github.com/tensorflow/tensorflow/pull/34660 otherwise it will conflict with the master once https://github.com/tensorflow/tensorflow/pull/34660 is merged. I am hoping github will take care of duplicated commits.

aaroey reviewed (commented):
Thanks Pooya for the fix. We'll need to wait for #34660 be merged first.

pooyadavoodi commented:
> Thanks Pooya for the fix. We'll need to wait for #34660 be merged first.

Thanks. I think merging them at the same time wouldn't hurt. git is smart enough to figure out the duplicates. Even if this PR gets merged first, the other PR would become an empty PR.

aaroey committed:
Merge branch 'master' into add_implicit_batch_experimental

aaroey committed:
Revert the change to gen_tftrt_model.py

Commit:
Add disable_non_trt_optimizers

This allows to create a rewriter_config that has all the non-TRT
optimizers off.

aaroey commented on tensorflow/python/compiler/tensorrt/trt_convert.py:
```diff
@@ -297,11 +300,31 @@ def get_tensorrt_rewriter_config(conversion_params, is_v2=False):
   rewriter_config_with_trt = rewriter_config_pb2.RewriterConfig()
-  if conversion_params.rewriter_config_template is None:
+
+  if not disable_non_trt_optimizers:
```
If template is not None do we still want to add these? How about:
```
if rewriter_config_template is None:
  ...extend(["constfold", ...])
  ...add trt optimizer
if disable_non_trt_optimizers:
  ...disable them
```

pooyadavoodi replied:
I was wondering about the same thing.
I am not sure if `rewriter_config_with_trt.CopyFrom` respects that.

You could say since `disable_non_trt_optimizers` sounds generic, it should also affect the template.

aaroey replied:
Yeah CopyFrom will overwrite that so we need to do that at the end of the method.

pooyadavoodi replied:
I just moved the disablement to the end of the function. Please review.

pooyadavoodi replied:
I merged the two if conditions.

Commit:
Move turnning off non-trt optimizers to after CopyFrom template

Disabling optimizers should happen after CopyFrom the temaplte
otherwise the template can overwrite the disablement.

Commit:
Move extending of non-TRT optimizers to inside next if condition

This is fine because CopyFrom of rewriter_config_template will overwrite
the disablement of optimizers and thus we can skip them.

gbaned commented:
@pooyadavoodi Could you please address Ubuntu Sanity errors? Thanks!

Commit:
Add use_implcit_batch to convert_nodes_test

Commit:
Delete unused import
